test/extensions/filters/http/dynamic_forward_proxy/proxy_filter_integration_test.cc

```diff
@@ -137,6 +144,32 @@ TEST_P(ProxyFilterIntegrationTest, RequestWithBody) {
   EXPECT_EQ(1, test_server_->counter("dns_cache.foo.host_added")->value());
 }
 
+TEST_P(ProxyFilterIntegrationTest, RequestWithPause) {
+  // Add a pause before dynamic forward proxy to regression test
+  // https://github.com/envoyproxy/envoy/issues/17514
+  buffer_ = true;
+  setup();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  const Http::TestRequestHeaderMapImpl request_headers{
+      {":method", "POST"},
+      {":path", "/test/long/url"},
+      {":scheme", "http"},
+      {":authority",
+       fmt::format("localhost:{}", fake_upstreams_[0]->localAddress()->ip()->port())}};
+
+  auto encoder_decoder = codec_client_->startRequest(request_headers);
+  auto& encoder = encoder_decoder.first;
+
+  // Send some data, but not the entire body.
+  std::string data(1024, 'a');
+  Buffer::OwnedImpl send1(data);
+  encoder.encodeData(send1, false);
+
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_FALSE(upstream_request_->waitForData(*dispatcher_, 1, std::chrono::milliseconds(100)));
+}
```

Comment:
This test isn't representative, because DNS is immediately resolved, and `dynamic-forward-proxy` resumes before processing body, so `buffer-body-filter` never had a chance to pause it:
```
decode headers called: filter=0x15b9fdbcae00 status=0  // buffer headers (continue)
decode headers called: filter=0x15b9fdbf1f10 status=4  // dfp headers (pause)
continuing filter chain: filter=0x15b9fdbf1f10         // dfp headers (continue)
decode headers called: filter=0x15b9fe1019d0 status=1  // router headers
decode data called: filter=0x15b9fdbcae00 status=1     // buffer body (pause)
```
Compare with the reported issue (https://github.com/envoyproxy/envoy/issues/17514#issuecomment-891616783):
```
decode headers called: filter=0x444afcea08c0 status=0  // wasm headers (continue)
decode headers called: filter=0x444afcea09a0 status=4  // dfp headers (pause)
decode data called: filter=0x444afcea08c0 status=1     // wasm body (pause)
decode data called: filter=0x444afcea08c0 status=1     // wasm body (pause)
continuing filter chain: filter=0x444afcea09a0         // dfp headers (continue)
decode headers called: filter=0x444afcea0a10 status=1  // router headers
decode data called: filter=0x444afcea09a0 status=0     // dfp body (continue)
decode data called: filter=0x444afcea0a10 status=3     // router body
```

```

Reply:
It looks like we may need a custom filter to test this case.